cocos/core/pipeline/ui/ui-stage.ts

```diff
@@ -91,6 +94,7 @@ export class UIStage extends RenderStage {
         cmdBuff.end();
 
         bufs[0] = cmdBuff;
-        this._device!.queue.submit(bufs);
+        this._device!.queue.submit(bufs, this._fence!);
+        this._fence!.wait(); this._fence!.reset();
```

Comment:
Why reset directly after wait ?

Reply:
The same question. Who notify it?

Reply:
the command queue will signale it after finishing all the work on GPU

Reply:
> The command queue will signale it after finishing all the work on GPU.

 If so, then i don't understand why `reset` is needed?

Reply:
`reset` is needed so that it can be signaled again

Reply:
Right, so I already removed that, fences should not be of concern now for any module outside GFX for now since they are all single-threaded. I was adding it then for a vulkan-specific validation error, but now I think we can/should just do it inside GFX to avoid the unnecessary synchonization hassle in a single-threaded context.

Reply:
But the fence structure will be useful in the future once we start digging around multi-threaded rendering though, so the majority of this work is still valid.